src/reader.rs

```diff
@@ -60,267 +68,439 @@ pub struct R1CSFile<F: PrimeField> {
-    // println!("wtns version {}", version);
+    let version = reader
+        .read_u32::<LittleEndian>()
+        .map_err(|err| ReadIntegerError { source: err.into() })?;
+
     if version > 2 {
-        bail!("unsupported file version");
+        return Err(WitnessVersionNotSupported(version.to_string()));
     }
-    let num_sections = reader.read_u32::<LittleEndian>()?;
+    let num_sections = reader
+        .read_u32::<LittleEndian>()
+        .map_err(|err| ReadIntegerError { source: err.into() })?;
     if num_sections != 2 {
-        bail!("invalid num sections");
+        return Err(SectionCountError(num_sections.to_string()));
     }
-    // read the first section
-    let sec_type = reader.read_u32::<LittleEndian>()?;
+    // Read the first section.
+    let sec_type = reader
+        .read_u32::<LittleEndian>()
+        .map_err(|err| ReadIntegerError { source: err.into() })?;
     if sec_type != 1 {
-        bail!("invalid section type");
+        return Err(SectionTypeError(1.to_string(), sec_type.to_string()));
     }
-    let sec_size = reader.read_u64::<LittleEndian>()?;
+    let sec_size = reader
+        .read_u64::<LittleEndian>()
+        .map_err(|err| ReadIntegerError { source: err.into() })?;
     if sec_size != 4 + 32 + 4 {
-        bail!("invalid section len")
+        return Err(SectionLengthError(
+            (4 + 32 + 4).to_string(),
+            sec_size.to_string(),
+        ));
     }
-    let field_size = reader.read_u32::<LittleEndian>()?;
+    let field_size = reader
+        .read_u32::<LittleEndian>()
+        .map_err(|err| ReadIntegerError { source: err.into() })?;
     if field_size != 32 {
-        bail!("invalid field byte size");
+        return Err(FieldByteSizeError(32.to_string(), field_size.to_string()));
     }
     let mut prime = vec![0u8; field_size as usize];
-    reader.read_exact(&mut prime)?;
-    // if prime != hex!("010000f093f5e1439170b97948e833285d588181b64550b829a031e1724e6430") {
-    //     bail!("invalid curve prime {:?}", prime);
-    // }
-    let witness_len = reader.read_u32::<LittleEndian>()?;
-    //println!("witness len {}", witness_len);
-    let sec_type = reader.read_u32::<LittleEndian>()?;
+    reader
+        .read_exact(&mut prime)
+        .map_err(|err| ReadBytesError { source: err.into() })?;
+
+    // Read the second section.
+    let witness_len = reader
+        .read_u32::<LittleEndian>()
+        .map_err(|err| ReadIntegerError { source: err.into() })?;
+    let sec_type = reader
+        .read_u32::<LittleEndian>()
+        .map_err(|err| ReadIntegerError { source: err.into() })?;
     if sec_type != 2 {
-        bail!("invalid section type");
+        return Err(SectionTypeError(2.to_string(), sec_type.to_string()));
     }
-    let sec_size = reader.read_u64::<LittleEndian>()?;
+    let sec_size = reader
+        .read_u64::<LittleEndian>()
+        .map_err(|err| ReadIntegerError { source: err.into() })?;
```

Comment:
I've actually been used to `#[source]` for error propagation. But it seems that if there are no information provided along side the source error `#[from]` is more than good enough. I'll implement this :+1:

Reply:
After giving it some thoughts it seems sticking with  `#[source]` is better in this case. 

We have multiple errors that originate from `io::Error`. As we can only have one variant annotated `#[from]` for one error type it would mean having to pass complementary information to it, [which should not be done](https://docs.rs/thiserror/latest/thiserror/).

As I'd like to keep our granularity, I think we should stick with `#[source]`

Reply:
OK!